### ISSUE TITLE
JPA - 3장 요약 정리

### ISSUE THREAD

**Issue by 2xel** [beadss/jpa-study#3]
# JPA 스터디

## ORM(Object Relational Mapping)

- 객체와 RDB 사이에 매핑을 정의한 메타데이터를 이용 Application 내의 객체를 RDB의 테이블로 자동으로 영속화 하는 프레임워크
- 영속화(Persistence)
  - [Persistence simply means that we would like our application’s data to outlive the applications process. In Java terms, we would like the state of (some of) our objects to live beyond the scope of the JVM so that the same state is available later.](http://hibernate.org/orm/what-is-an-orm/)
  - Java 에서 생각해보면 객체의 상태를 JVM의 범위를 넘어서 지속하여 추후에 동일한 상태를 이용하려는 것

## JPA(Java Persistence API)

- Java ORM 기술 표준
- Java EE 및 Java SE 환경 내에서 지속성 및 오브젝트/관계 맵핑 관리를 정의
- EJB에서 제공되던 엔터티 빈(Entity Bean)을 대체하는 기술이다. 이를 구현한 구현체로는 Hibernate, OpenJPA, EclipseLink 와 같은것들이 있고 이에 따른 표준 인터페이스

### @Entity

- JPA를 사용해서 테이블과 매핑할 클래스에 @Entity라는 선언을 해주면 이 클래스는 JPA가 관리해준다.
- name이라는 속성을 따로 설정하지 않으면 클래스 이름을 그대로 사용한다.

### @Table

- 엔티티 클래스에 매핑할 테이블 정보를 알려준다. 이 어노테이션을 생략하면 클래스 이름을 테이블 이름으로 매핑한다.
- 더 정확히는 엔티티 이름을 사용한다.

### @ID

- 엔티티 클래스의 필드를 테이블의 기본키(PK)에 매핑한다.

### @Column

- 필드를 컬럼에 매핑한다.
- 해당 Attribute와 매핑되는 컬럼정보를 입력하기 위한 것으로 Attribute 위에 기재한다.
- 컬럼명과 Attribute명이 일치할 경우 생략가능

# 3. 영속성 관리

### persistence.xml

- JPA의 메타정보를 입력 web.xml 처럼 스펙에 정해진 규격

```xml
    <persistence-unit name="jpabook">
        <properties>
            <property name="hibernate.diarect" value="${hibernate.dialect}" />
            <property name="hibernate.hbm2ddl.auto" value="${hibernate.hbm2ddl.auto}" />
            <property name="javax.persistence.jdbc.driver" value="${db.driver}" />
            <property name="javax.persistence.jdbc.url" value="${db.url}" />
            <property name="javax.persistence.jdbc.user" value="${db.user}" />
            <property name="javax.persistence.jdbc.password" value="${db.password}" />
        </properties>
    </persistence-unit>
```

### EntityManagerFactory

```java
EntityManagerFactory emf = Persistence.createEntityManagerFactory("jpabook");
```

- 앞서 작성한 persistence.xml의 내용을 바탕으로 엔티티 매니저 팩토리를 생성한다.
- 엔티티 매니저 팩토리는 엔티티 매니저를 만드는 공장이다.
- 엔티티 매니저 팩토리는 여러 스레드가 동시에 접근해도 안전하므로 서로 다른 쓰레드에서 공유가능

### EntityManager

```java
EntityManager em = emf.createEntityManager();
```

- 엔티티 매니저 팩토리를 이용해 엔티티 매니저를 생성한다.
- 엔티티 매니저는 엔티티를 관리하는 관리자다.
- 엔티티 매니저는 엔티티를 저장하고, 수정하고, 삭제하고, 조회하는 등 엔티티와 관련된 모든일을 처리한다.
- 개발자 입장에서는 엔티티 매니저는 엔티티를 저장하는 가상의 데이터베이스라고 생각하면 된다.
- 엔티티 매니저는 여러 스레드가 동시에 접근하면 동시성 문제가 발생하므로 스레드 간에 절대 공유하면 안된다.

### 영속성 컨텍스

- 물리적이 아닌 논리적인 개념
- 엔티티를 영구 저장하는 환경이라는 뜻
- 엔티티 매니저를 통해 영속성 컨텍스트에 접근, 관리 할 수 있다.
- 여러 엔티티 매니저가 같은 영속성 컨텍스트에 접근할 수도 있다.

## 엔티티의 생명주기

- 엔티티에는 4가지 상태가 존재
- 비영속(new/transient), 영속(managed), 준영속(detached), 삭제(removed)

### 비영속

- 엔티티 객체를 생성했지만 아직 영속성 컨텍스트와 관계가 없는 상태

### 영속

- 엔티티 매니저를 통해 엔티티를 영속성 컨텍스트에 저장한 상태
- 영속성 컨텍스트에 의해 관리된다라는 뜻
- em.find()나 JPQL을 사용해서 조회한 엔티티도 영속성 컨텍스트가 관리하는 영속 상태

### 준영속

- 영속성 컨텍스트가 관리하던 영속 상태의 엔티티를 영속성 컨텍트가 더이상 관리하지 않는 상태
- em.detach(), em.clear(), em.close()등을 호출해서 영속성 컨텍스트가 관리하던 엔티티를 준영속 상태로 만들수 있음

### 삭제

- 엔티티를 영속성 컨텍스트와 데이터베이스에서 삭제

## 영속성 컨텍스트의 특징

- 영속성 컨텍스트는 엔티티를 @Id로 테이블의 기본 키와 매핑한 값으로 구별한다.
- 따라서 영속 상태에는 반드시 식별자 값이 있어야 하며 없으면 예외가 발생
- 영속성 컨텍스트에 엔티티를 저장한 후 바로 데이테베이스에 저장하는 것이 아닌 flush()를 호출했을때 저장한다.

### 영속성 컨텍스트 사용시 장점

- 1차 캐시 기능을 한다.
  - 1차 캐시의 키는 식별자 값이다.
  - em.find()를 호출하면 먼저 1차 캐시에서 엔티티를 찾고 만약 없으면 데이터베이스에서 조회한다.
  - 데이터베이스에서 조회후 1차 캐시에 저장한 후에 영속상태의 엔티티를 반환한다.
  - 이후 한 트랜젝션안에서는 엔티티 인스턴스는 1차 캐시에 있으므로 이 엔티티를 조회시 메모리에 있는 1차 캐시에서 바로 불러오므로 성능상의 이점을 누릴 수 있음
- 동일성 보장
  - 동일성과 동등성 ==, equal
  - 영속성 컨텍스트에서 관리하는 엔티티 인스턴스는 동일성을 보장한다.
- 쓰기지연
  - 엔티티 매니저는 트랜잭션을 커밋하기 직전까지 데이터베이스에 엔티티를 저장하지 않고 내부 쿼리 저장소에 쿼리를 모아뒀다가 트랜잭션이 커밋할 때 모아둔 쿼리를 데이터베이스에 보낸다.
  - 플러시는 영속성 컨텍스트의 변경내용을 데이터베이스에 동기화하는 작업인데 커밋시 영속성 컨텍스트의 변경 내용을 데이터베이스와 동기화한 후 실제 데이터베이스의 트랜잭션을 커밋한다.
  - 이러한 쓰기 지연이 가능한 이유는 같은 트랜잭션 범위 안에서 실행되기 때문이다.
  - 이 기능을 잘 활용하면 모아둔 등록 쿼리를 데이터베이스에 한번에 전달해서 성능 최적화할 수 있다.
- 변경감지(dirty checking)

  - JPA는 엔티티를 영속성 컨텍스트에 보관할 때, 최초의 상태를 복사해서 저장해두는데 이것을 스냅샷이라 한다.
  - 플러시 시점에 스냅샷과 엔티티를 비교해서 변경된 엔티티를 찾고 변경된 엔티티가 있으면 수정 쿼리를 생성해서 지연 SQL 저장소에 보낸다.
  - 당연한 말이지만 변경감지는 영속성 컨텍스트가 관리하는 영속 상태의 엔티티에만 적용된다.
  - JPA의 기본전략은 업데이트시 엔티티의 모든 필드를 업데이트 하는데 이는 데이터의 전송량이 증가하는 단점이 있지만 수정쿼리가 항상 같다는 장점(수정쿠러리를 미리 생성해두고 재사용, 동일한 쿼리를 보내면 데이터베이스는 이전에 파싱된 쿼리를 재사용)으로 인해 모든 필드를 다 업데이트 한다.
  - 필드가 많거나 저장되는 내용이 너무 크면 수정된 데이터만 사용해서 동적으로 UPDATE SQL을 생성하는 전략을 선택하면 된다.(@DynamicUpdate)

## 플러시(flush)

- 변경감지 동작(엔티티를 스냅샷과 비교) -> 수정된 엔티티의 수정쿼리를 쓰기지연 SQL 저장소에 등록 -> 쓰기지연 SQL 저장소의 쿼리를 데이터베이스에 전송

  - 영속성 컨텍스트를 플러시 하는 방법 3가지

  1. em.flush() 호출
  2. 트랜젝션 커밋
  3. JPQL 쿼리 실행 시

- 직접호출은 테스트나 MyBatis등과 함께 사용할때를 제외하고는 거의 사용하지 않음
- 트랜잭션을 커밋하기 전에는 꼭 플러시를 호출해야 하는데 JPA는 기본적으로 트랜잭션을 커밋할 때 플러시를 자동으로 호출한다.
- JPQL은 SQL로 변환되어 실제 데이터베이스에서 엔티티를 조회하는데 영속성 컨텍스트에서 관리하는 엔티티들이 아직 데이터베이스에 반영되지 않았을때 문제가 발생할 수도 있으므로 JPQL을 실행하면 자동으로 플러시를 호출한다.
- 참고로 식별자 기준으로 조회하는 find() 메서드를 호출할때는 플러시가 실행되지 않는다.

### 플러시 모드 옵션

- FlushModeType.AUTO 커밋이나 쿼리를 실행할 때 플러시(기본값)
- FlushModeType.COMMIT 커밋할때만 플러시
- 대부분은 기본, 성능최적화가 필요할때 COMMIT 모드를 사용할 경우가 있음

## 준영속

- 영속성 컨텍스트가 관리하는 영속 상태의 엔티티가 영속성 컨텍스트에서 분리된 것을 준영속 상태라 한다.
- 준영속 상태의 엔티티는 영속성 컨텍스트가 제공하는 기능을 사용할 수 없다.
- 영속 상태의 엔티티를 준영속 상태로 만드는 방법은 3가지가 있다.
  - em.detach(entity) : 특정 엔티티만 준영속 상태로 전환
  - em.clear() : 영속성 컨텍스트를 초기화
  - em.close() : 영속성 컨텍스트를 종료

### detach()

- 영속성 컨텍스트에게 더 이상 해당 엔티티를 관리하지 말라는 것
- 메서드를 호출하는 순간 1차 캐시 + 쓰기 지연 저장소에 저장된 SQL에 저장된 모든 정보가 제거

### clear()

- detach(entity)는 특정 엔티티 하나만을 준영속 상태로 만들었다면 clear()는 영속성 컨텍스트를 초기화해 해당 영속성 컨텍스트의 모든 엔티티를 준영속 상태로 만드는 것
- 마찬가지로 1차 캐시 + 쓰기 지연 저장소에 저장된 모든 정보가 제거된다.

### close()

- 영속성 컨텍스트를 종료
- 영속성 컨텍스트가 종료되어 더 이상 영속성 컨텍스트가 존재하지 않음
- 개발자가 직접 close를 호출하여 준영속 상태로 만드는 일은 매우 드물다.

### 준영속 상태의 특징

- 영속성 컨텍스트가 관리하지 않으므로 영속성 컨텍스트가 제공하는 어떠한 기능도 동작하지 않는다.
- 비영속상태는 식별자값이 있을수도 있고 없을수도 있다.(물론 영속성컨텍스트가 관리하기 위해선 식별자가 없으면 오류가 남) 하지만 준영속 상태는 이미 한번 영속성 상태였으므로 반드시 식별자 값을 가지고 있다.
- 지연로딩이 불가능하다. 영속성 컨텍스트가 더 이상 관리하지 않으므로 영속성 컨텍스트가 실제 객체 대신 프록시 객체를 로딩해두고 해당 객체를 실제 사용할때 데이터를 불러오는 방법인 지연로딩시 문제가 발생한다.

### merge()

- 준영속 상태의 엔티티를 다시 영속 상태로 변경하는 방법
- 이때는 새로운 영속 상태의 엔티티를 반환한다.(==으로 비교가능 한지 봤으면 좋겠다.)

**Comment by beadss** [comment]
아래 내용 답변드립니다.
```
merge()
* 이때는 새로운 영속 상태의 엔티티를 반환한다.(==으로 비교가능 한지 봤으면 좋겠다.)
```
동일하지 않은 새로운 객체입니다.

**Comment by 2xel** [comment]
트랜젝션 격리수준에 대해서 토론해봐야할꺼같다.
처음 듣는 개념이다.
